### PR TITLE
putObjectPartial: Fix putobject partial to address anonymous and smal…

### PR DESCRIPTION
**Pull request by harshavardhana**
…l object cases.

### CODE REVIEW

File: api-put-object-partial.go [hunk 1]
```diff
@@ -34,6 +37,176 @@ func (c Client) PutObjectPartial(bucketName, objectName string, data ReadAtClose
...
+				Key:        objectName,
+				BucketName: bucketName,
+			}
+		}
+		// Do not compute MD5 for anonymous requests to Amazon S3. Uploads upto 5GiB in size.
+		n, err := c.putAnonymousPartial(bucketName, objectName, data, size, contentType)
+		return n, err
+	}
+
+	// Large file upload is initiated for uploads for input data size if its greater than 5MiB.
```
Comment (abperiasamy) [inline api-put-object-partial.go]: This should be default last case.

File: api-put-object-partial.go [hunk 2]
```diff
@@ -34,6 +37,176 @@ func (c Client) PutObjectPartial(bucketName, objectName string, data ReadAtClose
...
+		return n, err
+	}
+	n, err = c.putSmallObjectPartial(bucketName, objectName, data, size, contentType)
+	return n, err
+
+}
+
+// putNoChecksumPartial special function used Google Cloud Storage. This special function
+// is used for Google Cloud Storage since Google's multipart API is not S3 compatible.
+func (c Client) putNoChecksumPartial(bucketName, objectName string, data ReadAtCloser, size int64, contentType string) (n int64, err error) {
```
Comment (abperiasamy) [inline api-put-object-partial.go]: putNoChecksumPartial -> putPartialNoChecksum

File: api-put-object-partial.go [hunk 3]
```diff
@@ -34,6 +37,176 @@ func (c Client) PutObjectPartial(bucketName, objectName string, data ReadAtClose
...
+
+	// Create a new pipe to stage the reads.
+	reader, writer := io.Pipe()
+	var readAtOffset int64
+
+	// Initiate a routine to start writing.
+	go func() {
+		for {
+			// readAt defaults to reading at 5MiB buffer.
+			readAtBuffer := make([]byte, 1024*1024*5)
```
Comment (abperiasamy) [inline api-put-object-partial.go]: should be declared outside the loop once.

File: api-put-object-partial.go [hunk 4]
```diff
@@ -34,6 +37,176 @@ func (c Client) PutObjectPartial(bucketName, objectName string, data ReadAtClose
...
+	// For anonymous requests, we will not calculate sha256 and md5sum.
+	putObjMetadata := putObjectMetadata{
+		MD5Sum:      nil,
+		Sha256Sum:   nil,
+		ReadCloser:  reader,
+		Size:        size,
+		ContentType: contentType,
+	}
+	// Execute put object.
+	if _, err := c.putObject(bucketName, objectName, putObjMetadata); err != nil {
```
Comment (abperiasamy) [inline api-put-object-partial.go]: should never mask written len.

File: api-put-object-partial.go [hunk 5]
```diff
@@ -34,6 +37,176 @@ func (c Client) PutObjectPartial(bucketName, objectName string, data ReadAtClose
...
+			}
+			readAtOffset += int64(writeSize)
+			if rerr == io.EOF {
+				writer.Close()
+				return
+			}
+		}
+	}()
+	// For anonymous requests, we will not calculate sha256 and md5sum.
+	putObjMetadata := putObjectMetadata{
```
Comment (abperiasamy) [inline api-put-object-partial.go]: This is not the right categorization.

File: api-put-object-partial.go [hunk 6]
```diff
@@ -34,6 +37,176 @@ func (c Client) PutObjectPartial(bucketName, objectName string, data ReadAtClose
...
+		if err != io.EOF {
+			return 0, err
+		}
+	}
+	if int64(readAtSize) != size {
+		return 0, ErrUnexpectedEOF(int64(readAtSize), size, bucketName, objectName)
+	}
+
+	// Construct a new PUT object metadata.
+	putObjMetadata := putObjectMetadata{
```
Comment (abperiasamy) [inline api-put-object-partial.go]: putObjectData

**Comment by krishnasrinivas** [comment]
looks good

**Pull request merged**